File: bindings/python/examples/quickstart.py
"""One agent turn with a tool written in Python, printed as it happens.

    OPENAI_API_KEY=... uv run python examples/quickstart.py [model]

`model` defaults to `openai/gpt-5.4-mini`; its provider's API key has to be in the
environment (`OPENAI_API_KEY`, `ANTHROPIC_API_KEY`, ...).
"""

import asyncio
import json
import sys

import ailoy

QUERY = "Is it warmer in Seoul or in Busan right now?"


# A tool is any callable. The model's arguments arrive as keyword arguments, and what it
# returns is the result the model reads.
async def temperature(city: str) -> dict:
    await asyncio.sleep(0.1)  # stands in for a real lookup
    return {"city": city, "celsius": {"Seoul": 18.5, "Busan": 21.0}.get(city)}


async def main(model: str) -> None:
    desc = ailoy.register_tool(
        {
            "name": "temperature",
            "description": "The temperature in a city now, in Celsius.",
            "parameters": {
                "type": "object",
                "properties": {"city": {"type": "string", "description": "The city's name"}},
                "required": ["city"],
            },
        },
        temperature,
    )

    agent = await ailoy.AgentBuilder(model).instruction("Answer in one sentence.").tool(desc).build()

    async for output in agent.run(QUERY):
        message = output["message"]
        if message["role"] == "assistant":
            for part in message["contents"]:
                if part["type"] == "text":
                    print(part["text"])
            for call in message.get("tool_calls", []):
                function = call["function"]
                print(f"→ {function['name']} {json.dumps(function['arguments'])}")
        elif message["role"] == "tool":
            for part in message["contents"]:
                print(f"← {json.dumps(part.get('value'))}")


if __name__ == "__main__":
    asyncio.run(main(sys.argv[1] if len(sys.argv) > 1 else "openai/gpt-5.4-mini"))

File: bindings/node/examples/quickstart.mjs
// An agent with a tool of its own, run for one turn.
//
// The model is resolved from the default provider, which is seeded from the API keys in the
// environment — `ANTHROPIC_API_KEY` for the model named here.
//
//     ANTHROPIC_API_KEY=... node examples/quickstart.mjs

import { createRequire } from 'node:module'

const { AgentBuilder, registerTool } = createRequire(import.meta.url)('../index.js')

// A tool is its description and a function. The model's arguments arrive as one object; what
// the function returns, or resolves to, is the tool's result.
const weather = registerTool(
  {
    name: 'weather',
    description: 'The current weather in a city.',
    parameters: {
      type: 'object',
      properties: { city: { type: 'string' } },
      required: ['city'],
    },
  },
  async ({ city }) => ({ city, sky: 'clear', celsius: 21 }),
)

const agent = await new AgentBuilder('anthropic/claude-sonnet-5')
  .instruction('Answer in one sentence.')
  .tool(weather)
  .build()

try {
  // Deltas as the model writes them; each tool's result arrives whole.
  for await (const { delta } of agent.runStream('What is the weather in Seoul?')) {
    for (const part of delta.contents) {
      if (part.type === 'text') process.stdout.write(part.text)
    }
  }
  process.stdout.write('\n')
} finally {
  await agent.close()
}
